File: contract/fixtures/sample-repo/src/math.ts
import { clamp } from "./util";

// Multi-line signature: must collapse whitespace + trim before hashing.
export function add(a: number,
                    b: number): number {
  return a + b;
}

// Generics + default parameter value + varargs (rest param).
export function reduceSum<T>(items: T[],
                            seed: number = 0,
                            ...extra: number[]): number {
  let total = seed;
  for (const _ of items) {
    total = add(total, 1);
  }
  return clamp(total, 0, 100);
}

export interface Point {
  x: number;
  y: number;
}

export class Vec2 {
  constructor(public p: Point) {}

  length(): number {
    return add(this.p.x, this.p.y);
  }
}

File: contract/fixtures/sample-repo/src/math.go
// Go fixture: multi-line signature, variadic params, cross-symbol call.
package sample

// Multi-line signature + variadic param (Go's varargs).
func ReduceSum(
	seed int,
	items ...int,
) int {
	total := seed
	for range items {
		total = Add(total, 1)
	}
	return Clamp(total, 0, 100)
}

func Add(a int, b int) int {
	return a + b
}

func Clamp(value int, lo int, hi int) int {
	if value < lo {
		return lo
	}
	if value > hi {
		return hi
	}
	return value
}
